x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/DeleteStepTests.java

```diff
@@ -149,31 +149,27 @@ public void testPerformActionThrowsExceptionIfIndexIsTheDataStreamWriteIndex() {
             .metadata(Metadata.builder().put(index1, false).put(sourceIndexMetadata, false).put(dataStream).build())
             .build();
 
-        IllegalStateException illegalStateException = expectThrows(
-            IllegalStateException.class,
-            () -> createRandomInstance().performDuringNoSnapshot(sourceIndexMetadata, clusterState, new ActionListener<>() {
-                @Override
-                public void onResponse(Void complete) {
-                    fail("unexpected listener callback");
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    fail("unexpected listener callback");
-                }
-            })
-        );
-        assertThat(
-            illegalStateException.getMessage(),
-            is(
-                "index ["
-                    + sourceIndexMetadata.getIndex().getName()
-                    + "] is the write index for data stream ["
-                    + dataStreamName
-                    + "]. stopping execution of lifecycle [test-ilm-policy] as a data stream's write index cannot be deleted. "
-                    + "manually rolling over the index will resume the execution of the policy as the index will not be the "
-                    + "data stream's write index anymore"
-            )
-        );
+        createRandomInstance().performDuringNoSnapshot(sourceIndexMetadata, clusterState, new ActionListener<>() {
+            @Override
+            public void onResponse(Void complete) {
+                fail("unexpected listener callback");
+            }
+
+            @Override
+            public void onFailure(Exception e) {
```

Comment:
I think that now that this isn't throwing directly, we need to have a latch or other mechanism to ensure that the `onFailure` handler was actually invoked. Otherwise if we were to introduce a bug where neither `onResponse` nor `onFailure` were called, then we wouldn't hit any asserts and the test would pass (when it shouldn't).

Reply:
The execution in this particular case is not async as we're not getting to the callback as part of a client interaction - this is all sync as it's part of the step validation.
I've stubbed the client to fail in case it's being called in this test so that changes to the ILM step structure will yield this test to fail.

Reply:
Ah I (later) understood what you meant - opened https://github.com/elastic/elasticsearch/pull/105914 to make sure we fail the test if the listener is not called at all